best-first-cv/script.whizzml

```diff
@@ -0,0 +1,561 @@
+    (if (empty? seed)
+        options
+        (assoc options "seed" seed))))
+
+;; safe-delete
+;;
+;; deletes resources ignoring errors
+;; Inputs:
+;;   id: (resource-id) resource to be deleted
+;;
+;; Output: (boolean) true if successful, false if not
+;;
+(define (safe-delete id)
+  (try (delete id)
+       (catch e
+         (log-info (str "Error deleting resource " id " ignored"))
+         false)))
+
+
+
+;-----------------------------------------------------------------
+
+
+
+;; Do best-first feature selection.  Given a dataset and a target
+;; number of features iteratively construct models for each feature,
+;; evaluate them, and add the feature corresponding to the best
+;; cv to the running set of features.  Stop when you reach the
+;; target number, or you run out of features.
+(define (select-features dataset-id
+                         nfeatures
+                         objective-id
+                         options
+                         k-folds
+                         pre-selected)
+  (let (obj-id (get-objective dataset-id objective-id)
+        k-fold-datasets (create-k-folds dataset-id k-folds)
+        input-ids (default-inputs dataset-id obj-id)
+        important-ids (important-fields dataset-id obj-id)
+        _ (log-info important-ids)
+        potentials (make-complementary important-ids pre-selected))
+    (loop (selected pre-selected
+           potentials potentials
+           queue [])
+
+      (if (or (>= (count selected) nfeatures) (empty? potentials))
+        (response dataset-id k-fold-datasets selected queue)
+        (let (_ (log-info "Making new candidates...")
+              candidates (make-candidates selected
+                                          (map (lambda (x) (x "feature"))
+                                          potentials))
+              _ (log-info "Selecting feature..." candidates)
+              next-feat-info (select-feature k-fold-datasets
+                                             obj-id
+                                             options
+                                             candidates
+                                             potentials)
+              _ (log-info "Next selected feature: " (next-feat-info "feature")))
+          (recur (append selected (next-feat-info "feature"))
+                 (make-complementary potentials [(next-feat-info "feature")])
+                 (cons {"features" (append selected
+                                           (next-feat-info "feature"))
+                              "phi" (next-feat-info "phi")} queue)))))))
+
+;; A simple function to get the max value in a list
+(define (get-max xs) (reduce (lambda (x y) (if (> x y) x y)) (head xs) xs))
+
+
+;; Filters the features in one list to be complementary with the ones in
+;; another list
+(define (make-complementary y-list x-list)
+  (iterate (acc y-list x x-list)
+    (filter (lambda (y) (not (= x (y "feature")))) acc)))
```

Comment:
so x-list is  a list of features but y-list is a list of maps that have a "feature" key? if so, maybe we could say it explicitly in the comment... we are doing this, right?:

```
(let (fs (map (lambda (y) (y "feature")) y-list))
  (filter (lambda (x) (not (member? x fs))) x-list))
```

(that's just for my own understanding, feel free to keep your version if you find it clearer)